File: data-structures/go/src/hashmap/hash_map.go
package hashmap

import "fmt"

const (
	defaultCapacity = 16
	loadFactorLimit = 0.75
	fnvOffsetBasis  = 14695981039346656037
	fnvPrime        = 1099511628211
)

type entry[K comparable, V any] struct {
	key   K
	value V
	next  *entry[K, V]
}

type HashMap[K comparable, V any] struct {
	buckets  []*entry[K, V]
	size     int
	capacity int
}

func New[K comparable, V any]() *HashMap[K, V] {
	return &HashMap[K, V]{
		buckets:  make([]*entry[K, V], defaultCapacity),
		capacity: defaultCapacity,
	}
}

// fnvHash implements FNV-1a hash algorithm
func fnvHash(s string) uint64 {
	h := uint64(fnvOffsetBasis)
	for i := 0; i < len(s); i++ {
		h ^= uint64(s[i])
		h *= fnvPrime
	}
	return h
}

func (m *HashMap[K, V]) hash(key K) int {
	h := fnvHash(fmt.Sprintf("%v", key))
	return int(h % uint64(m.capacity))
}

func (m *HashMap[K, V]) Put(key K, value V) {
	if float64(m.size+1)/float64(m.capacity) > loadFactorLimit {
		m.rehash()
	}
	idx := m.hash(key)
	for e := m.buckets[idx]; e != nil; e = e.next {
		if e.key == key {
			e.value = value
			return
		}
	}
	m.buckets[idx] = &entry[K, V]{key: key, value: value, next: m.buckets[idx]}
	m.size++
}

func (m *HashMap[K, V]) Get(key K) (V, bool) {
	var zero V
	idx := m.hash(key)
	for e := m.buckets[idx]; e != nil; e = e.next {
		if e.key == key {
			return e.value, true
		}
	}
	return zero, false
}

func (m *HashMap[K, V]) Remove(key K) (V, bool) {
	var zero V
	idx := m.hash(key)
	var prev *entry[K, V]
	for e := m.buckets[idx]; e != nil; e = e.next {
		if e.key == key {
			if prev == nil {
				m.buckets[idx] = e.next
			} else {
				prev.next = e.next
			}
			m.size--
			return e.value, true
		}
		prev = e
	}
	return zero, false
}

func (m *HashMap[K, V]) Contains(key K) bool {
	_, found := m.Get(key)
	return found
}

func (m *HashMap[K, V]) Size() int {
	return m.size
}

func (m *HashMap[K, V]) IsEmpty() bool {
	return m.size == 0
}

func (m *HashMap[K, V]) Clear() {
	m.buckets = make([]*entry[K, V], m.capacity)
	m.size = 0
}

func (m *HashMap[K, V]) Keys() []K {
	keys := make([]K, 0, m.size)
	for _, bucket := range m.buckets {
		for e := bucket; e != nil; e = e.next {
			keys = append(keys, e.key)
		}
	}
	return keys
}

func (m *HashMap[K, V]) Values() []V {
	values := make([]V, 0, m.size)
	for _, bucket := range m.buckets {
		for e := bucket; e != nil; e = e.next {
			values = append(values, e.value)
		}
	}
	return values
}

func (m *HashMap[K, V]) Capacity() int {
	return m.capacity
}

// Clone creates a copy of this HashMap.
// Note: This performs a shallow copy of values. If values are mutable
// objects (pointers, slices, maps), modifications to them will be visible
// in both the original and cloned HashMap.
func (m *HashMap[K, V]) Clone() *HashMap[K, V] {
	clone := &HashMap[K, V]{
		buckets:  make([]*entry[K, V], m.capacity),
		size:     m.size,
		capacity: m.capacity,
	}
	for i, bucket := range m.buckets {
		var prev *entry[K, V]
		for e := bucket; e != nil; e = e.next {
			newEntry := &entry[K, V]{key: e.key, value: e.value}
			if prev == nil {
				clone.buckets[i] = newEntry
			} else {
				prev.next = newEntry
			}
			prev = newEntry
		}
	}
	return clone
}

func (m *HashMap[K, V]) rehash() {
	oldBuckets := m.buckets
	m.capacity *= 2
	m.buckets = make([]*entry[K, V], m.capacity)
	m.size = 0
	for _, bucket := range oldBuckets {
		for e := bucket; e != nil; e = e.next {
			m.Put(e.key, e.value)
		}
	}
}

File: data-structures/python/src/hash_map.py
class HashMap:
    def __init__(self, capacity=16):
        self._capacity = capacity
        self._size = 0
        self._buckets = [[] for _ in range(capacity)]

    def _bucket_index(self, key):
        return hash(key) % self._capacity

    def _rehash(self):
        old_buckets = self._buckets
        self._capacity *= 2
        self._buckets = [[] for _ in range(self._capacity)]
        self._size = 0
        for bucket in old_buckets:
            for key, value in bucket:
                self.put(key, value)

    def put(self, key, value):
        index = self._bucket_index(key)
        bucket = self._buckets[index]
        for i, (k, v) in enumerate(bucket):
            if k == key:
                bucket[i] = (key, value)
                return
        bucket.append((key, value))
        self._size += 1
        if self._size / self._capacity > 0.75:
            self._rehash()

    def get(self, key):
        index = self._bucket_index(key)
        for k, v in self._buckets[index]:
            if k == key:
                return v
        raise KeyError(key)

    def get_or(self, key, default):
        try:
            return self.get(key)
        except KeyError:
            return default

    def remove(self, key):
        index = self._bucket_index(key)
        bucket = self._buckets[index]
        for i, (k, v) in enumerate(bucket):
            if k == key:
                bucket.pop(i)
                self._size -= 1
                return v
        return None

    def contains(self, key):
        index = self._bucket_index(key)
        for k, v in self._buckets[index]:
            if k == key:
                return True
        return False

    def size(self):
        return self._size

    def is_empty(self):
        return self._size == 0

    def clear(self):
        self._buckets = [[] for _ in range(self._capacity)]
        self._size = 0

    def keys(self):
        result = []
        for bucket in self._buckets:
            for k, v in bucket:
                result.append(k)
        return result

    def values(self):
        result = []
        for bucket in self._buckets:
            for k, v in bucket:
                result.append(v)
        return result

    def copy(self):
        """Create a copy of this HashMap.

        Note: This performs a shallow copy of values. If values are mutable
        objects, modifications to them will be visible in both the original
        and copied HashMap.
        """
        new_map = HashMap(self._capacity)
        for bucket in self._buckets:
            for k, v in bucket:
                new_map.put(k, v)
        return new_map

    def __len__(self):
        return self._size

    def __contains__(self, key):
        return self.contains(key)

    def __getitem__(self, key):
        return self.get(key)

    def __setitem__(self, key, value):
        self.put(key, value)

    def __delitem__(self, key):
        if not self.contains(key):
            raise KeyError(key)
        self.remove(key)

    def __iter__(self):
        for bucket in self._buckets:
            for k, v in bucket:
                yield k
